drivers/can/can_renesas_ra.c

```diff
@@ -0,0 +1,1114 @@
+
+		/* Ignore AFL entry which will be changed */
+		reg->CFDGAFLIGNENT_b.IRN = afl_offset;
+		reg->CFDGAFLIGNCTR = ((CFDGAFLIGNCTR_KEY_CODE << R_CANFD_CFDGAFLIGNCTR_KEY_Pos) |
+				      BIT(R_CANFD_CFDGAFLIGNCTR_IREN_Pos));
+		reg->CFDGAFLECTR_b.AFLDAE = 1;
+
+		/* Change configurations for AFL entry */
+		reg->CFDGAFL[afl_offset] = *(R_CANFD_CFDGAFL_Type *)(afl);
+
+		/* Lock AFL entry access */
+		reg->CFDGAFLECTR_b.AFLDAE = 0;
+		reg->CFDGAFLIGNCTR = ((CFDGAFLIGNCTR_KEY_CODE << R_CANFD_CFDGAFLIGNCTR_KEY_Pos));
+	}
+}
+
+#if CONFIG_CAN_MANUAL_RECOVERY_MODE
+static int recover_bus(const struct device *dev, k_timeout_t timeout)
+{
+	struct can_renesas_ra_data *data = dev->data;
+	canfd_instance_ctrl_t *p_ctrl = data->fsp_can.p_ctrl;
+	R_CANFD_Type *reg = p_ctrl->p_reg;
+	uint32_t cfdcnctr = reg->CFDC->CTR;
+	int ret = 0;
+
+	if (reg->CFDC->ERFL_b.BOEF == 0) {
+		/* Channel bus-off entry not detected */
+		goto end;
+	}
+
+	reg->CFDC->CTR_b.BOM = 0x00; /* Switch to Normal Bus-Off mode (comply with ISO 11898-1) */
+	reg->CFDC->CTR_b.RTBO = 1;   /* Force channel state to return from bus-off */
+
+	int64_t start_ticks = k_uptime_ticks();
+
+	while (reg->CFDC->ERFL_b.BORF == 0) {
+		if ((k_uptime_ticks() - start_ticks) > timeout.ticks) {
+			ret = -EAGAIN;
+			goto end;
+		}
+	}
+end:
+	reg->CFDC->CTR = cfdcnctr; /* Restore channel configuration */
+	return ret;
+}
+#endif
+
+static inline void can_renesas_ra_call_tx_cb(const struct device *dev, int err)
+{
+	struct can_renesas_ra_data *data = dev->data;
+
+	can_tx_callback_t cb = data->tx_cb;
+
+	if (cb != NULL) {
+		data->tx_cb = NULL;
+		cb(dev, err, data->tx_usr_data);
+		k_sem_give(&data->tx_sem);
+	}
+}
+
+static inline void can_renesas_ra_call_rx_cb(const struct device *dev, can_callback_args_t *p_args)
+{
+	struct can_renesas_ra_data *data = dev->data;
+	struct can_renesas_ra_filter *rx_filter = data->rx_filter;
+	const struct can_renesas_ra_cfg *cfg = dev->config;
+	struct can_frame frame = {
+		.dlc = can_bytes_to_dlc(p_args->frame.data_length_code),
+		.id = p_args->frame.id,
+		.flags = (((p_args->frame.id_mode == CAN_ID_MODE_EXTENDED) ? CAN_FRAME_IDE : 0UL) |
+			  ((p_args->frame.type == CAN_FRAME_TYPE_REMOTE) ? CAN_FRAME_RTR : 0UL) |
+			  ((p_args->frame.options & CANFD_FRAME_OPTION_FD) != 0 ? CAN_FRAME_FDF
+										: 0UL) |
+			  ((p_args->frame.options & CANFD_FRAME_OPTION_ERROR) != 0 ? CAN_FRAME_ESI
+										   : 0UL) |
+			  ((p_args->frame.options & CANFD_FRAME_OPTION_BRS) != 0 ? CAN_FRAME_BRS
+										 : 0UL)),
+	};
+
+	memcpy(frame.data, p_args->frame.data, p_args->frame.data_length_code);
+
+	for (uint32_t i = 0; i < cfg->rx_filter_num; i++) {
+		if (NULL == rx_filter->rx_cb) {
+			rx_filter++;
+			continue; /* this filter has not set yet */
+		}
+
+		if (!can_frame_matches_filter(&frame, &rx_filter->filter)) {
+			rx_filter++;
+			continue; /* filter did not match */
+		}
+
+		rx_filter->rx_cb(dev, &frame, rx_filter->rx_usr_data);
+		break;
```

Comment:
This piece of code suffers from TOCTOU (https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) on the `rx_filter->rx_cb` function. It could have been changes between the check and the use.

Reply:
Can we just ignore this issue for now so we can merge this PR? Updating to replace with hardware filters can take a huge workload. I will create another PR to update for this.

Reply:
That's not what I am suggesting. I just meant you need a local copy of `rx_filter->rx_cb`, check _that_ against `NULL` and call _that_ for the function call. Otherwise the callback function value may have changed to NULL in between the check and the use.

That being said, a follow-up PR with hardware filter support would be very welcome.

Reply:
I fixed it